### PR TITLE
adds feature gates

### PR DESCRIPTION
**Pull request by p0lyn0mial**
**What this PR does / why we need it**: features can be enabled/disabled by setting the new command line flag, for example setting `-feature-gates=PrometheusEndpoint=true` exposes cluster's metrics HTTP endpoint. Note that the changes introduced in this pull depreciate 'enable-prometheus-endpoint' flag.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
#2268 

**Release note**:

```release-note
[ACTION REQUIRED] added a new command line flag to API server that accepts a set of key=value pairs that enables/disables various features. Existing `enable-prometheus-endpoint` flag is deprecated, the users should use `-feature-gates=PrometheusEndpoint=true` instead. 
```

### CODE REVIEW

File: api/hack/run-api.sh [hunk 1]
```diff
@@ -9,7 +9,7 @@ make -C $(dirname $0)/.. kubermatic-api
 
 KUBERMATIC_WORKERNAME=${KUBERMATIC_WORKERNAME:-$(uname -n)}
 
-# Please make sure to set -enable-prometheus-endpoint=true if you want to use that endpoint.
+# Please make sure to set -feature-gates=PrometheusEndpoint=true if you want to use that endpoint.
```
Comment (p0lyn0mial) [inline api/hack/run-api.sh:12]: @mrIncompetent  haven't seen `enable-prometheus-endpoint` flag being used in our code base, thus decided to depreciate it.

File: api/cmd/kubermatic-api/main.go [hunk 2]
```diff
@@ -161,9 +162,8 @@ func createAuthenticator(options serverRunOptions) (handler.Authenticator, error
 }
 
 func createAPIHandler(options serverRunOptions, prov providers, authenticator handler.Authenticator, updateManager *version.Manager) (http.HandlerFunc, error) {
-	// Only enable the metrics endpoint when prometheusEndpoint is true
 	var prometheusClient prometheusapi.Client
-	if options.prometheusEndpoint {
+	if options.featureGates.enabled(PrometheusEndpoint) {
```
Comment (p0lyn0mial) [inline api/cmd/kubermatic-api/main.go:166]: essence of this pull.

File: api/cmd/kubermatic-api/options.go [hunk 3]
```diff
@@ -38,9 +38,15 @@ func newServerRunOptions() serverRunOptions {
 	flag.StringVar(&s.tokenIssuer, "token-issuer", "", "URL of the OpenID token issuer. Example: http://auth.int.kubermatic.io")
 	flag.BoolVar(&s.tokenIssuerSkipTLSVerify, "token-issuer-skip-tls-verify", false, "SKip TLS verification for the token issuer")
 	flag.StringVar(&s.clientID, "client-id", "", "OpenID client ID")
+	flag.StringVar(&rawFeatureGates, "feature-gates", "", "A set of key=value pairs that describe feature gates for various features.")
```
Comment (p0lyn0mial) [inline api/cmd/kubermatic-api/options.go:41]: one flag for setting various features.

**Review by maciaszczykm — APPROVED** [review]

**Comment by mrIncompetent** [comment]
@p0lyn0mial Can you add this to the kubermatic controller-manager as well?

**Comment by p0lyn0mial** [comment]
> @p0lyn0mial Can you add this to the kubermatic controller-manager as well?

"feature gates" concept ? If so then I could create a dedicated package which could be used by api and controller.

**Comment by mrIncompetent** [comment]
Yes

**Comment by p0lyn0mial** [comment]
ok

File: api/hack/run-api.sh (hunk 1 above)
Comment (p0lyn0mial) [reply api/hack/run-api.sh:12]: @mrIncompetent I hope it it safe to remove the flag ?

Comment (mrIncompetent) [reply api/hack/run-api.sh:12]: sgtm, if you can add a `[ACTION REQUIRED]` prefix to the release note

**Pull request merged**